File: src/main/java/com/github/yuppieflu/stats/service/bucket/StatBucket.java
package com.github.yuppieflu.stats.service.bucket;

import com.github.yuppieflu.stats.service.domain.Measurement;
import lombok.RequiredArgsConstructor;

import java.util.concurrent.locks.ReadWriteLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;

@RequiredArgsConstructor
class StatBucket {
    private final int index;

    private final ReadWriteLock readWriteLock = new ReentrantReadWriteLock(true);
    private long count;
    private double max;
    private double min;
    private double sum;
    private long lastAddedTimestamp;

    void addMeasurement(Measurement m) {
        readWriteLock.writeLock().lock();
        try {
            if (lastAddedTimestamp - m.getTimestamp() > BucketsStorageService.ONE_BUCKET_MILLIS) {
                return;
            }
            if (m.getTimestamp() - lastAddedTimestamp > BucketsStorageService.ONE_BUCKET_MILLIS) {
                count = 1;
                max = m.getValue();
                min = m.getValue();
                sum = m.getValue();
            } else {
                count++;
                max = Math.max(max, m.getValue());
                min = Math.min(min, m.getValue());
                sum += m.getValue();
            }
            lastAddedTimestamp = m.getTimestamp();
        } finally {
            readWriteLock.writeLock().unlock();
        }
    }

    boolean shouldSkip(long timestamp) {
        readWriteLock.readLock().lock();
        try {
            if (count == 0) {
                return true;
            }
            int currentIndex = TimeUtils.secondOfMinute(timestamp);
            if (currentIndex == index) {
                return timestamp - lastAddedTimestamp > BucketsStorageService.ONE_BUCKET_MILLIS;
            }
            return timestamp - lastAddedTimestamp > BucketsStorageService.BUCKETS_SIZE_MILLIS;
        } finally {
            readWriteLock.readLock().unlock();
        }
    }

    StatBucketContent content() {
        readWriteLock.readLock().lock();
        try {
            return StatBucketContent.builder()
                                    .count(this.count)
                                    .max(this.max)
                                    .min(this.min)
                                    .sum(this.sum)
                                    .build();
        } finally {
            readWriteLock.readLock().unlock();
        }
    }
}

File: src/main/kotlin/com/github/yuppieflu/stats/service/bucket/StatBucket.kt
package com.github.yuppieflu.stats.service.bucket

import com.github.yuppieflu.stats.service.domain.Measurement
import java.util.concurrent.locks.ReentrantReadWriteLock
import kotlin.math.max
import kotlin.math.min

class StatBucket {
    private val readWriteLock = ReentrantReadWriteLock(true)
    private var count = 0L
    private var min = 0.0
    private var max = 0.0
    private var sum = 0.0
    private var lastAddedTimestamp = 0L

    fun addMeasurement(m: Measurement) {
        readWriteLock.writeLock().lock()
        try {
            if (lastAddedTimestamp < m.timestamp) {
                replaceOldData(m)
            } else {
                add(m)
            }
            lastAddedTimestamp = m.timestamp
        } finally {
            readWriteLock.writeLock().unlock()
        }
    }

    private fun replaceOldData(m: Measurement) {
        count = 1L
        min = m.value
        max = m.value
        sum = m.value
    }

    private fun add(m: Measurement) {
        count++
        min = min(min, m.value)
        max = max(max, m.value)
        sum += m.value
    }

    fun isWithin60SecondsFrom(timestamp: Long): Boolean {
        readWriteLock.readLock().lock()
        try {
            return timestamp - this.lastAddedTimestamp <= BucketsStorageService.SIZE
        } finally {
            readWriteLock.readLock().unlock()
        }
    }

    fun content(): StatBucketContent {
        readWriteLock.readLock().lock()
        try {
            return StatBucketContent(count = count, min = min, max = max, sum = sum)
        } finally {
            readWriteLock.readLock().unlock()
        }
    }
}